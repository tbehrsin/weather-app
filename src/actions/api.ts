
import * as constants from "../constants";

export interface IRequestStart {
  type: constants.api.REQUEST_START;
  key: string;
  options: IRequestOptions;
}

export interface IRequestError {
  type: constants.api.REQUEST_ERROR;
  key: string;
  error: Error;
}

export interface IRequestEnd {
  type: constants.api.REQUEST_END;
  key: string;
  body: any;
}

export type Action = IRequestStart | IRequestError | IRequestEnd;

export interface IRequestOptions {
  method?: string;
  url: string;
  query?: any;
  body?: any;
  bodyParser?: (body: any) => any;
}

export const request = (key: string, options: IRequestOptions): IRequestStart => ({
  key,
  options,
  type: constants.api.REQUEST_START,
});

export const error = (key: string, e: Error): IRequestError => ({
  error: e,
  key,
  type: constants.api.REQUEST_ERROR,
});

export const response = (key: string, body: any): IRequestEnd => ({
  body,
  key,
  type: constants.api.REQUEST_END,
});
