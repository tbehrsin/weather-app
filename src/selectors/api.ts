
import { createSelector } from "reselect";

import * as constants from "../constants";
import { RootState } from "../store";

export const domain = (state: RootState) => state.api;

export const isPending = createSelector(
  domain,
  (api: any) => (key: string) => !api.getIn(["requests", key, "body"]) && !api.getIn(["requests", key, "error"]),
);

export const isSuccess = createSelector(
  domain,
  (api: any) => (key: string) => !!api.getIn(["requests", key, "body"]),
);

export const isFail = createSelector(
  domain,
  (api: any) => (key: string) => !!api.getIn(["requests", key, "error"]),
);

export const requestKeys = createSelector(
  domain,
  (api: any) => api.get("requests").filter((r: any) => r.get("state") === constants.api.STATE_COMPLETE).keySeq().toJS(),
);

export const body = createSelector(
  domain,
  (api: any) => (key: string) => {
    const b = api.getIn(["requests", key, "body"]);

    if (b) {
      return b.toJS();
    }
  },
);

export const requestState = createSelector(
  domain,
  (api: any) => (key: string) => api.getIn(["requests", key, "state"]),
);

export const requestStateAll = createSelector(
  domain,
  (api: any) => {
    const error = !!api.get("requests").find((r: any) => r.get("state") === constants.api.STATE_ERROR);
    if (error) {
      return constants.api.STATE_ERROR;
    }

    const requesting = !!api.get("requests").find((r: any) => r.get("state") === constants.api.STATE_REQUESTING);
    if (requesting) {
      return constants.api.STATE_REQUESTING;
    }

    return constants.api.STATE_COMPLETE;
  },
);

export const errorAny = createSelector(
  domain,
  (api: any) => {
    const error = api.get("requests").find((r: any) => r.get("state") === constants.api.STATE_ERROR);
    if (error) {
      return error.toJS();
    }
  },
);
