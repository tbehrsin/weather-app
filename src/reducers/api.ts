
import { fromJS } from "immutable";

import * as actions from "../actions";
import * as constants from "../constants";

const INITIAL_STATE = {
  requests: {},
};

export default (state = fromJS(INITIAL_STATE), action: actions.api.Action = null) => {
  switch (action.type) {
  case constants.api.REQUEST_START: {
    const { key } = action;
    return state
      .setIn(["requests", key, "state"], constants.api.STATE_REQUESTING)
      .deleteIn(["requests", key, "refresh"]);
  }
  case constants.api.REQUEST_ERROR: {
    const { key, error } = action;
    return state
      .setIn(["requests", key, "error"], fromJS(error))
      .setIn(["requests", key, "state"], constants.api.STATE_ERROR)
      .deleteIn(["requests", key, "refresh"]);
  }
  case constants.api.REQUEST_END: {
    const { key, body } = action;
    return state
      .setIn(["requests", key, "body"], fromJS(body))
      .setIn(["requests", key, "state"], constants.api.STATE_COMPLETE);
  }
  default:
    return state;
  }
};
