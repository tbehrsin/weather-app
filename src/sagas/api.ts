
import querystring from "querystring";
import { put, takeEvery } from "redux-saga/effects";

import * as actions from "../actions";
import * as constants from "../constants";

function* request(action: actions.api.IRequestStart) {
  const { key, options: { method = "GET", url, query = null, bodyParser = (body: any) => body } } = action;

  try {
    const qs = querystring.stringify(query);

    const response = yield fetch(`${url}${qs ? `?${qs}` : ""}`, {
      method,
    });

    if (!response.ok) {
      throw new Error("request failed");
    }

    const json = yield response.json();
    const responseBody = bodyParser(json);

    yield put(actions.api.response(key, responseBody));
  } catch (error) {
    yield put(actions.api.error(key, error));
  }
}

export function* watchRequestStart() {
  yield takeEvery(constants.api.REQUEST_START, request);
}
