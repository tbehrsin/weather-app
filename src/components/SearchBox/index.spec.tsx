import { shallow, render } from "enzyme";
import React from "react";

import { SearchBox } from "./index";

import * as api from "../../actions/api";

jest.mock("../../actions/api");

describe("SearchBox", () => {
  it("should match snapshot", () => {
    const wrapper = render(<SearchBox placeholder="Test" />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should send request on submit", () => {

    const request = jest.fn().mockImplementationOnce(() => null);

    const event = { preventDefault: () => { /* NOOP */ }};
    const wrapper = shallow(<SearchBox placeholder="Test" request={request} />);
    wrapper.setState({ value: "Manchester, GB" });
    expect(wrapper.find("form").length).toBe(1);
    wrapper.find("form").simulate("submit", event);
    expect(request.mock.calls).toMatchSnapshot();
  });
});
