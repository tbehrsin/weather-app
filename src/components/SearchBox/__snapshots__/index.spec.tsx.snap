// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox should match snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "container",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "autofocus": "",
          "placeholder": "Test",
          "type": "text",
          "value": "",
        },
        "children": Array [],
        "name": "input",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "type": "submit",
          },
          "children": Array [
            Object {
              "data": "Search",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "button",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "type": undefined,
          },
          "x-attribsPrefix": Object {
            "type": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "autofocus": undefined,
          "placeholder": undefined,
          "type": undefined,
          "value": undefined,
        },
        "x-attribsPrefix": Object {
          "autofocus": undefined,
          "placeholder": undefined,
          "type": undefined,
          "value": undefined,
        },
      },
      Object {
        "attribs": Object {
          "type": "submit",
        },
        "children": Array [
          Object {
            "data": "Search",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "button",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "autofocus": "",
            "placeholder": "Test",
            "type": "text",
            "value": "",
          },
          "children": Array [],
          "name": "input",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "autofocus": undefined,
            "placeholder": undefined,
            "type": undefined,
            "value": undefined,
          },
          "x-attribsPrefix": Object {
            "autofocus": undefined,
            "placeholder": undefined,
            "type": undefined,
            "value": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "type": undefined,
        },
        "x-attribsPrefix": Object {
          "type": undefined,
        },
      },
    ],
    "name": "form",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`SearchBox should send request on submit 1`] = `
Array [
  Array [
    "weather-manchester-gb",
    Object {
      "query": Object {
        "APPID": "9715d5915c4a1408053c752f5af7bee6",
        "q": "manchester,gb",
        "units": "metric",
      },
      "url": "http://api.openweathermap.org/data/2.5/forecast",
    },
  ],
]
`;
